prime-router/src/test/kotlin/fhirengine/azure/FhirFunctionTests.kt

```diff
@@ -271,7 +405,334 @@ class FhirFunctionTests {
             fhirEngine.doWork(any(), any(), any())
             actionHistory.trackActionParams(queueMessage) // string
             actionHistory.trackLogs(emptyList()) // list actionLog
-            workflowEngine.recordAction(any())
+            workflowEngine.recordAction(any(), any())
+        }
+        verify(exactly = 0) {
+            queueMock.sendMessage(any(), any())
+        }
+    }
+
+    @Nested
+    @ExtendWith(ReportStreamTestDatabaseSetupExtension::class)
+    inner class FhirFunctionIntegrationTests {
+
+        private fun seedTask(
+            fileFormat: Report.Format,
+            nextAction: TaskAction,
+            nextEventAction: Event.EventAction
+        ): Report {
+            val report = Report(
+                fileFormat,
+                listOf(ClientSource(organization = oneOrganization.name, client = "Test Sender")),
+                1,
+                metadata = UnitTestUtils.simpleMetadata,
+                nextAction = nextAction,
+                topic = Topic.FULL_ELR,
+            )
+            ReportStreamTestDatabaseContainer.testDatabaseAccess.transact { txn ->
+                val action = Action()
+                val actionId = ReportStreamTestDatabaseContainer.testDatabaseAccess.insertAction(txn, action)
+                val reportFile = ReportFile().setSchemaTopic(Topic.FULL_ELR).setReportId(report.id)
+                    .setActionId(actionId).setSchemaName("").setBodyFormat(fileFormat.toString()).setItemCount(1)
+                ReportStreamTestDatabaseContainer.testDatabaseAccess.insertReportFile(
+                    reportFile, txn, action
+                )
+                ReportStreamTestDatabaseContainer.testDatabaseAccess.insertTask(
+                    report,
+                    fileFormat.toString().lowercase(),
+                    "http://${report.id}.${fileFormat.toString().lowercase()}",
+                    nextAction = ProcessEvent(
+                        nextEventAction,
+                        report.id,
+                        Options.None,
+                        emptyMap(),
+                        emptyList()
+                    ),
+                    txn
+                )
+            }
+
+            return report
+        }
+
+        @Test
+        fun `test does not update the DB or send messages on an error`() {
+            val report = seedTask(Report.Format.HL7, TaskAction.convert, Event.EventAction.CONVERT)
+
+            mockkObject(BlobAccess.Companion)
+            mockkObject(Message.Companion)
+            every { BlobAccess.Companion.downloadBlob(any()) } returns hl7_record.toByteArray()
+            every {
+                BlobAccess.Companion.uploadBody(
+                    any(),
+                    any(),
+                    any(),
+                    any(),
+                    any()
+                )
+            } throws RuntimeException("manual error")
```

Comment:
Force the transaction to rollback.  I actually think there was a bug in the code as written where rows could get written even if there was an error.